src/uct/cuda/cuda_ipc/cuda_ipc_md.c

```diff
@@ -50,21 +50,92 @@ uct_cuda_ipc_mem_add_reg(void *addr, uct_cuda_ipc_memh_t *memh,
 {
     uct_cuda_ipc_lkey_t *key;
     ucs_status_t status;
+#if HAVE_CUDA_FABRIC
+    CUmemAllocationHandleType fabric_type = CU_MEM_HANDLE_TYPE_FABRIC;
+    CUmemGenericAllocationHandle handle;
+    CUmemoryPool mempool;
+    int allowed_handle_types;
+#endif
 
     key = ucs_malloc(sizeof(*key), "uct_cuda_ipc_lkey_t");
     if (key == NULL) {
         return UCS_ERR_NO_MEMORY;
     }
 
+    memset(key, 0, sizeof(*key));
+
+    UCT_CUDADRV_FUNC_LOG_ERR(cuMemGetAddressRange(&key->d_bptr, &key->b_len,
+                (CUdeviceptr)addr));
+
+#if HAVE_CUDA_FABRIC
+    /* cuda_ipc can handle VMM, mallocasync, and legacy pinned device so need to
+     * pack appropriate handle */
+    status =
+        UCT_CUDADRV_FUNC(cuMemRetainAllocationHandle(&handle, addr), UCS_LOG_LEVEL_DIAG);
+    if (status != UCS_OK) {
+        /* Try legacy IPC first */
+        status =
+            UCT_CUDADRV_FUNC_LOG_DEBUG(cuIpcGetMemHandle(&key->ph.handle.legacy,
+                        (CUdeviceptr)addr));
+        if (status == UCS_OK) {
+            key->ph.handle_type = UCT_CUDA_IPC_KEY_HANDLE_TYPE_LEGACY;
+            ucs_trace("packed legacy handle for %p", addr);
+        } else {
+            status = UCT_CUDADRV_FUNC_LOG_ERR(cuPointerGetAttribute(&mempool,
+                        CU_POINTER_ATTRIBUTE_MEMPOOL_HANDLE, (CUdeviceptr)addr));
+            if (status == UCS_OK) {
+                status = UCT_CUDADRV_FUNC_LOG_ERR(cuPointerGetAttribute(
+                            &allowed_handle_types,
+                            CU_POINTER_ATTRIBUTE_ALLOWED_HANDLE_TYPES,
+                            (CUdeviceptr)addr));
+                if ((status != UCS_OK) ||
+                    (!(allowed_handle_types | fabric_type))) {
+                    status = UCS_ERR_NO_RESOURCE;
+                    goto err;
+                }
```

Comment:
why is this needed?

Reply:
Because user may allocate memory pool without fabric capabilities in which case cuda-ipc UCT cannot support fabric handle based IPC.